tls-client/HelloHttpsClient.cpp

```diff
@@ -49,216 +84,302 @@ HelloHttpsClient::HelloHttpsClient( const char *in_server_name,
     mbedtls_ssl_config_init(&ssl_conf);
 }
 
-HelloHttpsClient::~HelloHttpsClient( void )
+HelloHttpsClient::~HelloHttpsClient()
 {
     mbedtls_entropy_free(&entropy);
     mbedtls_ctr_drbg_free(&ctr_drbg);
     mbedtls_x509_crt_free(&cacert);
     mbedtls_ssl_free(&ssl);
     mbedtls_ssl_config_free(&ssl_conf);
 
-    if( eth_iface != NULL )
-        delete eth_iface;
-    if( socket != NULL )
-        delete socket;
+    socket.close();
 }
 
-int HelloHttpsClient::createTCPSocket( void )
+int HelloHttpsClient::run()
 {
     int ret;
-    const char *ip_addr;
+    size_t req_len, req_offset, resp_offset;
+    uint32_t flags;
+    bool resp_200, resp_hello;
+
+    /* Configure the EthernetInterface and TCPSocket */
+    if ((ret = configureTCPSocket()) != 0)
+        return ret;
+
+    /* Configure already initialized mbed TLS structures */
+    if ((ret = configureTlsContexts()) != 0)
+        return ret;
 
-    eth_iface = new EthernetInterface();
-    if( eth_iface == NULL )
-    {
-        mbedtls_printf( "Failed to allocate EthernetInterface\r\n" );
-        return( -1 );
+    /* Start a connection to the server */
+    if ((ret = socket.connect(server_name, server_port)) != NSAPI_ERROR_OK) {
+        mbedtls_printf("socket.connect() returned %d\r\n", ret);
+        return ret;
+    } else {
+        mbedtls_printf("Successfully connected to %s at port %u\r\n",
+                       server_name, server_port);
     }
 
-    /* Initialise the ethernet interface and start up the stack */
-    if( ( ret = eth_iface->connect() ) != 0 )
-    {
-        mbedtls_printf( "Failed call to eth_iface.connect(): %d\r\n", ret );
-        return( ret );
+    /* Start the TLS handshake */
+    mbedtls_printf("Starting the TLS handshake...\r\n");
+    do {
+        ret = mbedtls_ssl_handshake(&ssl);
+    } while(ret != 0 &&
+            (ret == MBEDTLS_ERR_SSL_WANT_READ ||
+            ret == MBEDTLS_ERR_SSL_WANT_WRITE));
+    if (ret < 0) {
+        mbedtls_printf("mbedtls_ssl_handshake() returned -0x%04X\r\n", -ret);
+        return ret;
+    } else {
+        mbedtls_printf("Successfully completed the TLS handshake\r\n");
     }
 
-    ip_addr = eth_iface->get_ip_address();
-    if( ip_addr != NULL )
-    {
-        mbedtls_printf( "Client IP address is %s\r\n", ip_addr );
+    /* Fill the request buffer */
+    ret = snprintf(gp_buf, GENERAL_PURPOSE_BUFFER_LENGTH - 1,
+                   "GET %s HTTP/1.1\nHost: %s\n\n", HTTP_REQUEST_FILE_PATH,
+                   server_name);
+    req_len = static_cast<size_t>(ret);
```

Comment:
This cast is redundant - it will be performed during the assignment

Reply:
It is cast implicitly, but I added the explicit cast to ensure that no signed to unsigned implicit cast errors are generated by the compiler...